File: 10-05-2024_Combination_Sum_II.cpp
class Solution{
public:

    void help(vector<int> &arr,int n,int k , int i , int sum, vector<int> &target , set<vector<int>> &ans)
    {
        if(sum==k)
        {
            ans.insert(target);
            return;
        }
        if(i==n)
        return;
        help(arr,n,k,i+1,sum,target,ans);
        if(sum+arr[i]<=k)
        {
            target.push_back(arr[i]);
            help(arr,n,k,i+1,sum+arr[i],target,ans);
            target.pop_back();
        }
    }
    vector<vector<int>> CombinationSum2(vector<int> arr,int n,int k)
    {
        sort(arr.begin(),arr.end());
        vector<int> target;
        set<vector<int>> ans;
        help(arr, n , k , 0 , 0 , target , ans);
        vector<vector<int>> result(ans.begin(),ans.end());
        return result;
    }
};

File: 10-05-2024_Combination_Sum_II.py
from typing import List

class Solution:
    def combinationSum2(self, candidates: List[int], target: int) -> List[List[int]]:
        def help(arr: List[int], n: int, k: int, i: int, sum_: int, target_: List[int], ans: set[tuple[int]]) -> None:
            if sum_ == k:
                ans.add(tuple(target_))
                return
            if i == n:
                return
            help(arr, n, k, i + 1, sum_, target_, ans)
            if sum_ + arr[i] <= k:
                target_.append(arr[i])
                help(arr, n, k, i + 1, sum_ + arr[i], target_, ans)
                target_.pop()

        candidates.sort()
        target_ = []
        ans = set()
        help(candidates, len(candidates), target, 0, 0, target_, ans)
        return list(ans)

